bundles/org.openhab.binding.flume/src/main/java/org/openhab/binding/flume/internal/handler/FlumeBridgeHandler.java

```diff
@@ -0,0 +1,303 @@
+            handleApiException(e);
+            return;
+        }
+
+        scheduler.execute(() -> {
+            if (!refreshDevices(true)) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                        "@text/offline.cloud-configuration-error");
+                return;
+            }
+
+            int pollingPeriod = Math.min(config.refreshIntervalCumulative, config.refreshIntervalInstantaneous);
+            pollingJob = scheduler.scheduleWithFixedDelay(this::pollDevices, 0,
+                    Duration.ofMinutes(pollingPeriod).toSeconds(), TimeUnit.SECONDS);
+            updateStatus(ThingStatus.ONLINE);
+        });
+    }
+
+    @Nullable
+    public List<FlumeApiDevice> apiListDevicesAction() {
+        try {
+            return api.getDeviceList();
+        } catch (FlumeApiException | IOException | InterruptedException | TimeoutException | ExecutionException e) {
+            handleApiException(e);
+            return null;
+        }
+    }
+
+    /*
+     * Will update the listDevicesCache if expired or forcedUpdate. Will iterate through the list and
+     * either notify of discovery to discoveryService or, if the device is already configured, will update
+     * the device info.
+     */
+    public boolean refreshDevices(boolean forcedUpdate) {
+        final FlumeDiscoveryService discovery = discoveryService;
+
+        if (forcedUpdate) {
+            apiListDevicesCache.invalidateValue();
+        }
+        @Nullable
+        List<FlumeApiDevice> listDevices = apiListDevicesCache.getValue();
+
+        if (listDevices == null) {
+            return false;
+        }
+
+        for (FlumeApiDevice dev : listDevices) {
+            if (dev.type == 2 && discovery != null) {
+                FlumeDeviceHandler deviceHandler = getFlumeDeviceHandler(dev.id);
+
+                if (deviceHandler == null) {
+                    // output ID of discovered device to log once to identify ID so it can be used for textual
+                    // configuration
+                    if (!logOnce) {
+                        logger.info("Flume Meter Device Discovered: ID: {}", dev.id);
+                        logOnce = true;
+                    }
+                    discovery.notifyDiscoveryDevice(dev.id);
+                } else {
+                    deviceHandler.updateDeviceInfo(dev);
+                }
+            }
+        }
+
+        return true;
+    }
+
+    @Nullable
+    public FlumeDeviceHandler getFlumeDeviceHandler(String id) {
+        List<Thing> things = getThing().getThings();
+
+        for (Thing t : things) {
+            if (!t.getThingTypeUID().equals(THING_TYPE_DEVICE)) {
+                continue;
+            }
+
+            FlumeDeviceHandler handler = (FlumeDeviceHandler) t.getHandler();
+            if (handler != null && handler.getId().equals(id)) {
+                return handler;
```

Comment:
Prevent double casting.
```suggestion
            if (t.getHandler() instanceof FlumeDeviceHandler deviceHandler && deviceHandler.getId().equals(id)) {
                return deviceHandler;
```

Reply:
I rewrote this using streams - so please check this again.